### PR TITLE
Implement ml_meth as an union.

### PR DESCRIPTION
**Pull request by mejrs**
Implements https://github.com/PyO3/pyo3/issues/2129

### CODE REVIEW

File: pyo3-ffi/src/methodobject.rs [hunk 1]
```diff
@@ -71,11 +78,44 @@ extern "C" {
...
+/// otherwise the behavior is undefined.
+///
+/// See the [Python C API documentation][1] for more information.
+///
+/// [1]: https://docs.python.org/3/c-api/structures.html#implementing-functions-and-methods
+#[repr(C)]
+#[derive(Copy, Clone)]
+pub union MlMeth {
+    /// This variant corresponds with [`METH_VARARGS`] *or* [`METH_NOARGS`] *or* [`METH_O`].
+    pub PyCFunction: Option<PyCFunction>,
```
Comment (mejrs) [inline pyo3-ffi/src/methodobject.rs:98]: Can these things ever be null? Are the options necessary here?

Comment (birkenfeld) [reply pyo3-ffi/src/methodobject.rs:98]: I don't think they can, a method slot always needs an implementation.

File: pyo3-ffi/src/methodobject.rs [hunk 2]
```diff
@@ -71,11 +78,44 @@ extern "C" {
...
+///
+/// This union must be accompanied by the correct [ml_flags](PyMethodDef::ml_flags),
+/// otherwise the behavior is undefined.
+///
+/// See the [Python C API documentation][1] for more information.
+///
+/// [1]: https://docs.python.org/3/c-api/structures.html#implementing-functions-and-methods
+#[repr(C)]
+#[derive(Copy, Clone)]
+pub union MlMeth {
```
Comment (birkenfeld) [inline pyo3-ffi/src/methodobject.rs:96]: I don't particularly like the name. Can we call it `PyMethodDefPointer` (similar to a union we already have without equivalent in CPython, `PyUnicodeObjectData`)?

File: pyo3-ffi/src/methodobject.rs (hunk 1 above)
Comment (davidhewitt) [reply pyo3-ffi/src/methodobject.rs:98]: I think so too. Note that strictly speaking C function pointers are nullable whereas Rust ones aren't, and `pyo3-ffi` handles this slightly weirdly by making the typedefs non-nullable and then wrapping them in `Option` in every struct /  function where they're used.

Here I think it's probably ok to drop the `Option`.

**Comment by davidhewitt** [comment]
Thanks, this does look a bit nicer than the transmutes. I think this needs a `CHANGELOG` entry!

File: pyo3-ffi/src/methodobject.rs [hunk 3]
```diff
@@ -71,17 +78,45 @@ extern "C" {
...
-    pub ml_meth: Option<PyCFunction>,
+    pub ml_meth: PyMethodDefPointer,
     pub ml_flags: c_int,
     pub ml_doc: *const c_char,
 }
 
-impl Default for PyMethodDef {
-    fn default() -> PyMethodDef {
-        unsafe { mem::zeroed() }
-    }
```
Comment (mejrs) [inline pyo3-ffi/src/methodobject.rs:82]: I don't think this default impl makes sense if PyMethodDefPointer is never null.

Comment (davidhewitt) [reply pyo3-ffi/src/methodobject.rs:82]: Agreed; I also actually generally don't like having any implementations of traits or methods for `pyo3-ffi` types if possible (except for `Copy` & `Clone`). Makes it closer to the C that they come from.

**Review by davidhewitt — APPROVED** [review]
LGTM, thanks!

**Pull request merged**